DiscordBot/Commands/ReactionBuffer.cs

```diff
@@ -0,0 +1,81 @@
+﻿using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using DiscordBot.DataAccess;
+using DiscordBot.DataAccess.Models;
+using DSharpPlus;
+
+namespace DiscordBot.Commands
+{
+    internal enum ReactionBufferState
+    {
+        Ready,
+        Collecting
+    }
+
+    internal class ReactionBuffer
+    {
+        private const int bufferDelayMilliseconds = 10 * 1000;
+
+        private readonly BlockingCollection<ResponseReaction> buffer;
+        private readonly SemaphoreSlim bufferStateLock;
+        private ReactionBufferState state;
+
+        private readonly IEventsSheetsService eventsSheetsService;
+        private readonly DiscordClient client;
+
+        public ReactionBuffer(IEventsSheetsService eventsSheetsService, DiscordClient client)
```

Comment:
I feel like this complication should be protected from the front end code if possible. I would expect the behaviour to be something like:
1. Users react to the message.
2. The reactions immediately appear on the message.
3. The reactions immediately get passed to the sheets service.
4. The sheets service only updates the google sheet in batches.

Since as far as I understand we're having to do this in batches due to the google sheets api restrictions. So it seems like it's a concern which should be hidden from the front end. What do you think? 🙂

Reply:
The reason we did this was so that it could perform the Discord updates that need to happen after the reactions have been processed (deleting the reactions off the message, and updating the messages with the new signups). I do agree that it feels like we're mixing concerns, but I'm not sure if we can fully separate them due to the back-and-forth nature of operations between the two services. Do you have any ideas? 🙂

Reply:
Deleting/updating the reactions off the message seems like it could happen straight away without needing to wait for the sheet to be updated. What exactly do you mean by updating the messages with the new signups? 🙂

Reply:
That's referring to the Discord message being updated to show the correct participants and their responses. This can only be done after the sheet has been updated we can only get updated information from the sheet.

Reply:
Yeah updating reactions happen first. 

The message for each event has a list of responses, and needs to be prodded to tell it to get the new version list and update the message. We don't want to do this as the reactions come in, because this would complicate how we process adding and clearing reactions together. Instead we can just decide that when the batch is processed, and the message just fetches the result. Sorry that wasn't clear in the first message 🙂

Reply:
Where exactly is this message and how is it being used? 🙂

Reply:
The message that the bot sends for the signups.
It shows the list of participants and their responses.
![image](https://user-images.githubusercontent.com/36454654/99665737-f2490c80-2a61-11eb-86c1-cb875fb8b1f6.png)

Reply:
So this is the message which people react to? And it gets updated when people respond? Is that really necessary? 🙂

Reply:
Yes that is the message that people react to to respond to events. We think it's a really nice thing to have: by putting the event responses in this message normal users can see who has signed up already in a simple and neat way. This might be useful for more complex events, where the type of signup may depend on other signups (eg. positions for a football game).

Currently there isn't another good way of getting this information to general users, who don't have access to the underlying Google Sheet, and don't have permission to be posting commands in the bot-command channel. We would rather not add them to this channel, because that would begin to break down the divide between the bot-admin roles and the normal users. What do you think? 🙂

Reply:
Can't you see who added the reaction by hovering over it like you can in slack?

Reply:
At the moment the reactions are being deleted as part of processing the reaction. This is a fairly common practice with Discord bots; it allows the bot to track which work has and hasn't been processed (eg. allowing it to catch up on work that was missed when the bot was turned off). We haven't implemented this yet but with this design it would not result in any behaviour change if we were to implement this.

The process at the moment roughly is: 1) react to a message 2) process the consequences of this reaction 3) remove the reaction, notify the user they have signed up, and update the message.

Part of the reason we chose to do this was that subscribing to reaction removals was much harder in the previous version of DSharpPlus, but could be easier now. We also vaguely remember talking about this in the meeting last Friday with Gareth, but we can't quite remember what was said. We could change it over to having persistent reactions for responses if you prefer 🙂

Reply:
I think that if updating the message is making the implementation significantly more complicated (which it looks like it is), then it's worth reconsidering whether it's something we need to do. It's not something I would necessarily expect a bot like this to do. You could check with Gareth what he thinks 🙂